source/blog/_posts/2018-08-23-against-microservices.md

```diff
@@ -0,0 +1,136 @@
+
+While this abstract definition might be a little hard to grasp, there is a very good example that showcases the effect your organizational structure has on the software you produce:
+
+>  "If you have four groups working on a compiler, you'll get a 4-pass compiler." - E. S. Raymond
+
+... because communication is most intense inside the team and gets sparser and sparser the further away another team is organization wise. Microservices are the logical conclusion to this. There needs to be a formal way to negotiate API interfaces between teams and that's it. This actually solves a real problem in Software development: **Scale** If you ever started a green field project with a projection for more then two developers and a initial runtime of more then a year you should know how hard it is to get everyone working. Huge Silicon Valley corporations want to start projects with hundreds of developers or maintain flagship products with with even more people. Communication becomes an overhead that delays a product significantly think of it like an expression of the big upfront design antipattern. The more people you have that need to work the costlier it becomes  to negotiate a single decision.  
+
+By giving the highest degree of freedom to your teams the structure of your software changes, too. Dependencies in such a huge product can no longer be top down tree like organized but are in fact more like a **bee hive**. Microservices can potentially be even more then that. The hive is multinational! Your services can be written in any language, come from any source and as long as network communication is possible you are fine. In extreme cases there can and will be C code communicating happily with a PHP script that notifies a Python app that provides data for Node app. Every one of these apps is self contained and uses a different set of tools for its build, deployment, integration and monitoring. Yeah!
+
+Now imagine something is broken. How can you inspect, debug and fix any issue in such a system? You need conventions!
+
+> "With great power comes great responsibility" - B. Parker
+
+If you want to base each new service on different technology you will find that your Microservices have multiple different solutions to the same common problem. [*Not invented here*](https://en.wikipedia.org/wiki/Not_invented_here) is the buildin result from this approach to team organizations.
+
+To keep with the bee hive metaphor: Even they grow over time from just a founding queen.
+
+In the end the freedom simply does not come for free. Where does scale outweigh central technical decision management? Some huge companies that develop concurrently on a single product of course have good reasons to go into that direction. And in the wild you will find scaled back solutions (e.g. single programming language, central logging and monitoring, centrally enforced architecture, even single process) that try to get the pros without the cons. But what is left in these cases?
```

Comment:
"In the end the freedom simply does not come for free" => "In the end this freedom isn't free "